File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head==null){
            return null;
        }
        ListNode first=head;
        ListNode second=head;
        ListNode remover=null;
        int delay = n;
        while(first!=null){
            if(delay<=0){
                remover=second;
                second = second.next;
            }
            first=first.next;
            delay--;
        }
        if(remover==null){
            return second.next;
        }
        remover.next = second.next;
        second.next = null;
        return head;
    }
}

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        #Stall the back pointer for n number of times
        #If at this point fast is on None:
        #return back.next
        #Then move both back and front pointer
        #Keep a previous pointer that points to front of back
        
        back = head
        front = head
        previous = None

        while(n > 0):
            n -= 1
            front = front.next
        
        if front == None:
            return back.next
        while(front!= None):
            front = front.next
            previous = back
            back = back.next
        
        previous.next = previous.next.next

        return head